File: example/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/example/shell/TreeQueryShell.java
package org.babyfish.jimmer.sql.example.shell;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectWriter;
import org.babyfish.jimmer.sql.SqlClient;
import org.babyfish.jimmer.sql.example.model.TreeNode;
import org.babyfish.jimmer.sql.example.model.TreeNodeFetcher;
import org.babyfish.jimmer.sql.example.model.TreeNodeTable;
import org.springframework.shell.standard.ShellComponent;
import org.springframework.shell.standard.ShellMethod;
import org.springframework.shell.standard.ShellOption;

import java.util.List;

@ShellComponent
public class TreeQueryShell {

    private final SqlClient sqlClient;

    private final ObjectWriter prettyWriter;

    public TreeQueryShell(SqlClient sqlClient, ObjectWriter prettyWriter) {
        this.sqlClient = sqlClient;
        this.prettyWriter = prettyWriter;
    }

    @ShellMethod(
            "Find treeNodes, can optionally exclude child nodes of one node" +
            "(Example: trees --no-recursive clothing)"
    )
    public void trees(
            @ShellOption(defaultValue = "") String noRecursive
    ) throws JsonProcessingException {

        List<TreeNode> rootNodes = sqlClient
                .createQuery(TreeNodeTable.class, (q, treeNode) -> {
                    q
                            .where(treeNode.parent().isNull())
                            .orderBy(treeNode.name());
                    return q.select(
                            treeNode.fetch(
                                    TreeNodeFetcher.$
                                            .allScalarFields()
                                            .childNodes(
                                                    TreeNodeFetcher.$
                                                            .allScalarFields(),
                                                    it -> it
                                                            .filter(args -> args.orderBy(args.getTable().name()))
                                                            .recursive((node, depth) ->
                                                                    noRecursive.isEmpty() ||
                                                                            !noRecursive.equalsIgnoreCase(node.name())
                                                            )
                                            )
                            )
                    );
                })
                .execute();

        for (TreeNode rootNode : rootNodes) {
            System.out.println(
                    prettyWriter.writeValueAsString(rootNode)
            );
        }
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/sql/shell/TreeQueryShell.kt
package org.babyfish.jimmer.example.kt.sql.shell

import com.fasterxml.jackson.databind.ObjectWriter
import org.babyfish.jimmer.example.kt.sql.model.TreeNode
import org.babyfish.jimmer.example.kt.sql.model.fetchBy
import org.babyfish.jimmer.example.kt.sql.model.name
import org.babyfish.jimmer.example.kt.sql.model.parent
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.ilike
import org.babyfish.jimmer.sql.kt.ast.table.isNull
import org.springframework.shell.standard.ShellComponent
import org.springframework.shell.standard.ShellMethod
import org.springframework.shell.standard.ShellOption

@ShellComponent
class TreeQueryShell(
    private val sqlClient: KSqlClient,
    private val prettyWriter: ObjectWriter
) {

    @ShellMethod(
        "Find treeNodes, can optionally exclude child nodes of some nodes" +
            "(Example: tree --no-recursive clothing,drinks)"
    )
    fun trees(
        @ShellOption(defaultValue = "") rootName: String,
        @ShellOption(defaultValue = "") noRecursive: String
    ) {
        val noRecursiveNames =
            if (noRecursive.isEmpty()) {
                emptySet()
            } else {
                noRecursive
                    .trim()
                    .split("\\s*,\\s*")
                    .map {
                        it.lowercase()
                    }
                    .toSet()
            }
        val rootNodes = sqlClient.createQuery(TreeNode::class) {
            where(table.parent.isNull())
            rootName.takeIf { it.isNotEmpty() }?.let {
                where(table.name ilike it)
            }
            orderBy(table.name)
            select(
                table.fetchBy {
                    allScalarFields()
                    childNodes({
                        recursive {
                            !noRecursiveNames.contains(
                                entity.name.lowercase()
                            )
                        }
                        filter {
                            orderBy(table.name)
                        }
                    }) {
                        allScalarFields()
                    }
                }
            )
        }.execute()

        for (rootNode in rootNodes) {
            println(prettyWriter.writeValueAsString(rootNode))
        }
    }
}